hi-ml-azure/src/health_azure/datasets.py

```diff
@@ -24,45 +30,185 @@ def get_datastore(workspace: Workspace, datastore_name: str) -> Datastore:
         return datastore
-    if datastore_name in datastores:
-        return datastores[datastore_name]
-    raise ValueError(f"Datastore \"{datastore_name}\" was not found in the \"{workspace.name}\" workspace. "
-                     f"Existing datastores: {existing_stores}")
+
+    datastores = workspace.datastores
+    if isinstance(datastores, DatastoreOperations):
+        return _retrieve_v2_datastore(datastores, datastore_name)
+    elif isinstance(datastores, dict):
+        return _retrieve_v1_datastore(datastores, datastore_name)
+    else:
+        raise ValueError(f"Unrecognised type for datastores: {type(datastores)}")
+
+
+def _retrieve_v1_dataset(dataset_name: str, workspace: Workspace) -> Optional[FileDataset]:
+    """
+    Retrieve an Azure ML v1 Dataset if it exists, otherwise return None
+
+    :param dataset_name: The name of the Dataset to look for.
+    :param workspace: An Azure ML Workspace object for retrieving the Dataset.
+    :return: A Dataset object if it is found, else None.
+    """
+    logging.info(f"Trying to retrieve AzureML Dataset '{dataset_name}'")
+    azureml_dataset = Dataset.get_by_name(workspace, name=dataset_name)
+    return azureml_dataset
 
 
-def get_or_create_dataset(workspace: Workspace, datastore_name: str, dataset_name: str) -> FileDataset:
+def _create_v1_dataset(datastore_name: str, dataset_name: str, workspace: Union[Workspace, MLClient]
+                       ) -> FileDataset:
+    logging.info(f"Retrieving datastore '{datastore_name}' from AzureML workspace")
+    # Ensure that a v1 workspace is used
+    workspace = get_workspace(aml_workspace=workspace)
+    datastore = get_datastore(workspace, datastore_name)
+    logging.info(f"Creating a new dataset from data in folder '{dataset_name}' in the datastore")
+    # Ensure that there is a / at the end of the file path, otherwise folder that share a prefix could create
+    # trouble (for example, folders foo and foo_bar exist, and I'm trying to create a dataset from "foo")
+    azureml_dataset = Dataset.File.from_files(path=(datastore, dataset_name + "/"))
+    logging.info("Registering the dataset for future use.")
+    azureml_dataset.register(workspace, name=dataset_name)
+    return azureml_dataset
+
+
+def _get_or_create_v1_dataset(datastore_name: str, dataset_name: str, workspace: Workspace) -> Dataset:
+    """
+    Attempt to retrieve a v1 Dataset object and return that, otherwise atempts to create and register
+    a v1 Dataset and return that.
+
+    :param datastore_name: The name of the Datastore to either retrieve or create and register the Dataset in.
+    :param dataset_name: The name of the Dataset to be retrieved or registered.
+    :param workspace: An Azure ML Workspace object.
+    :return: An Azure ML Dataset object with the provided dataset name, in the provided datastore.
+    """
+    try:
+        azureml_dataset = _retrieve_v1_dataset(dataset_name, workspace)
+        assert azureml_dataset is not None
```

Comment:
This is just for mypy, because the return type of  `_retrieve_v1_dataset` is Optional, but the return type of `_get_or_create_v1_dataset` should be a dataset (because if one can't be found it should be created). Note that we don't have the same issue for the v2 version of the function because there if the data can't be found it raises an Exception rather than returning None

Reply:
Can we then check if it's None and raise an exception? It seems like it can cause some issues otherwise. It is now caught by the assertion, but maybe worth raising an exception rather than an assertion Error if not found?

Reply:
Just realised this is wrong, it shouldn't be wrapped in a `try/except`. If `_retrieve_v1_dataset` returns None, then we should then call `_create_v1_dataset`. Have updated it now

Reply:
Sorry  was wrong above - it does throw an exception. So I have added the try/catch back in